numpy/_core/src/umath/loops_hyperbolic.dispatch.cpp.src

```diff
@@ -381,29 +487,58 @@ simd_tanh_f64(const double *src, npy_intp ssrc, double *dst, npy_intp sdst, npy_
 
 #if NPY_SIMD_F32
 
+struct hwy_f32x2 {
+    vec_f32 val[2];
+};
 
-    // For architectures without efficient gather / scatter instructions, it is
-    // better to use a transposed LUT where we can load all coefficients for an
-    // index linearly.  In order to keep the same vertical calculation, we
-    // transpose the coef. into lanes.  A 4x4 transpose is all that's
-    // supported so we require `npyv_nlanes_f32` == 4.
-    #if npyv_nlanes_f32 == 4
-    #define TANHF_TRANSPOSED_LUT
-    // Define missing universal intrinsics used below
-    #if !defined(npyv_get_lane_u32)
-        #if defined(NPY_HAVE_ASIMD)
-            #define UNDEF_npyv_get_lane_u32
-            #define npyv_get_lane_u32 vgetq_lane_u32
-        #elif defined(NPY_HAVE_SSE41)
-            #define UNDEF_npyv_get_lane_u32
-            #define npyv_get_lane_u32 _mm_extract_epi32
-        #else
-            #undef TANHF_TRANSPOSED_LUT
-        #endif
-    #endif // !defined(npyv_get_lane_u32)
-    #endif // npyv_nlanes_f32 == 4
-
-static void
+HWY_ATTR NPY_FINLINE hwy_f32x2 zip_f32(vec_f32 a, vec_f32 b){
+    hwy_f32x2 res;
+    res.val[0] = hn::InterleaveLower(f32, a, b);
+    res.val[1] = hn::InterleaveUpper(f32, a, b);
+    return res;
+}
+
+HWY_ATTR NPY_FINLINE vec_f32 lut_32_f32(const float * lut, vec_u32 idx){
+    if constexpr(hn::Lanes(f32) == 16){
+        const vec_f32 lut0 = hn::Load(f32, lut);
+        const vec_f32 lut1 = hn::Load(f32, lut + 16);
+        return hn::TwoTablesLookupLanes(f32, lut0, lut1, hn::IndicesFromVec(f32, idx));
+    }else if constexpr (hn::Lanes(f32) == 8){
+        const vec_f32 lut0 = hn::Load(f32, lut);
+        const vec_f32 lut1 = hn::Load(f32, lut + 8);
+        const vec_f32 lut2 = hn::Load(f32, lut + 16);
+        const vec_f32 lut3 = hn::Load(f32, lut + 24);
+        
+        const auto high_mask = hn::Ne(hn::ShiftRight<4>(idx), hn::Zero(u32));
+        const auto load_mask = hn::And(idx, hn::Set(u32, 0b1111));
+        
+        const vec_f32 lut_low = hn::TwoTablesLookupLanes(f32, lut0, lut1, hn::IndicesFromVec(f32, load_mask));
+        const vec_f32 lut_high = hn::TwoTablesLookupLanes(f32, lut2, lut3, hn::IndicesFromVec(f32, load_mask));
+        
+        return hn::IfThenElse(hn::RebindMask(f32, high_mask), lut_high, lut_low);
+    }else{
+        return hn::GatherIndex(f32, lut, hn::BitCast(s32, idx));
+    }
+}
+
+// For architectures without efficient gather / scatter instructions, it is
+// better to use a transposed LUT where we can load all coefficients for an
+// index linearly.  In order to keep the same vertical calculation, we
+// transpose the coef. into lanes.  A 4x4 transpose is all that's
+// supported so we require `npyv_nlanes_f32` == 4.
+#if npyv_nlanes_f32 == 4
```

Comment:
As written, the HWY_TARGET condition means SSE4 or better. Lanes also isn't constexpr for SVE, so I think we'd have to check MaxLanes(f32) <= 4. Maybe we can drop the target check, because avoiding gather would be a win on any platform, right?

Reply:
I've made this change; I changed the target condition to `HWY_TARGET >= HWY_SSE4` which I think is correct, so AVX2 and AVX512 don't use the transposed LUTs. It'd be great if someone could double check that that's correct.